New task

zuwarka opened the pull request.

zuwarka committed:
Create README.md

zuwarka committed:
Update README.md

zuwarka committed:
Update README.md

zuwarka committed:
Create Node.rb

zuwarka committed:
Create LinkedList.rb

zuwarka committed:
Create PaginationHelper.rb

zuwarka committed:
Create main.rb

В данном файле идет проверка прописанных классов. Для второго задания реализовано три примера: первый пример - из условия; второй пример - свой на нечетное количество элементов массива и четное количество элементов на страницу; третий пример - отрицательное число для элементов на страницу.

Alex6Alex commented on homework05 (18.07.2022)/LinkedList.rb:
```diff
@@ -0,0 +1,93 @@
+
+  def find(data)
+    node = @head
+    return false unless node.next_node
```
Лучше здесь вернуть nil - объекта нет. Чтобы при вызове данного метода не приходилось учитывать, что объект у нас будет либо Node, либо false. Nil хоть тоже в руби отдельный класс, но больше подходит при ситуации, когда чего-то нет

Alex6Alex commented on homework05 (18.07.2022)/LinkedList.rb:
```diff
@@ -0,0 +1,93 @@
+    end
+  end
+
+  def find_before(data)
```
Так как, по условию задания, этот метод не был указан, лучше определить его как приватный или защищенный

Alex6Alex commented on homework05 (18.07.2022)/Node.rb:
```diff
@@ -0,0 +1,11 @@
+  end
+
+  def to_s
+    "#{@data}"
```
Эта запись равна `@data.to_s`, поэтому лучше использовать ее. Тогда это еще будет больше походить на делегацию: мы вызываем метод to_s узла, а он в свою очередь просто переводит выполнение на метод to_s объекта @data

Alex6Alex commented on homework05 (18.07.2022)/PaginationHelper.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def to_s
+    "#{@m_array}"
```
Здесь тоже бы подошло `@m_array.to_s`

Alex6Alex commented on homework05 (18.07.2022)/PaginationHelper.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def page_count
+    puts "There are #{(@m_array.length.to_f / @item_on_page.to_f).ceil} pages"
```
По условию, метод должен просто вернуть число, не предложение.

Alex6Alex commented on homework05 (18.07.2022)/PaginationHelper.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def item_count
+    puts "There are #{@m_array.length} items"
```
Здесь тоже лучше вернуть просто число

Alex6Alex commented on homework05 (18.07.2022)/PaginationHelper.rb:
```diff
@@ -0,0 +1,64 @@
+    puts "There are #{@m_array.length} items"
+  end
+
+  def page_count_for_class
```
Похоже на дубликат page_count в плане вычислений

Alex6Alex commented on homework05 (18.07.2022)/PaginationHelper.rb:
```diff
@@ -0,0 +1,64 @@
+  def page_item_count(page_number)
+    if (page_number < page_count_for_class) && (page_number >= 0)
+      if page_number == page_count_for_class - 1
+        puts "There are #{@m_array.length - ((@m_array.length / item_on_page).floor * item_on_page)} on the page"
```
Здесь тоже стоит убрать вывод строк

Alex6Alex commented on homework05 (18.07.2022)/PaginationHelper.rb:
```diff
@@ -0,0 +1,64 @@
+  def page_index(item_index)
+    index = @m_array.length - item_index
+    if index <= 0
+      puts "No such index, -1"
```
И здесь)

Alex6Alex commented on homework05 (18.07.2022)/PaginationHelper.rb:
```diff
@@ -0,0 +1,64 @@
+# helper.page_index(-10) # -1
+
+class PaginationHelper
+  attr_accessor :m_array, :item_on_page
```
Не похоже, чтобы данные accessor использовались. В методах класса используются переменные, а извне доступ вряд ли к этим переменным понадобится. Менять их после инициализации класса тоже не задумывалось

zuwarka committed:
Update main.rb

Добавлены puts в методы PaginationHelper

zuwarka committed:
Update PaginationHelper.rb

Исправлены недочеты с возвращаемыми данными методами

zuwarka committed:
Update Node.rb

Делегирован метод to_s

zuwarka committed:
Update LinkedList.rb

Добавлены поясняющие комментарии к методу find; запривачены методы find_after и find_before

zuwarka committed:
Update LinkedList.rb

zuwarka committed:
Update LinkedList.rb

Отредактирована 38 строка на return nil unless node.next_node

Alex6Alex commented on homework05 (18.07.2022)/main.rb:
```diff
@@ -0,0 +1,69 @@
+require_relative 'LinkedList'
```
Файлы с руби кодом тоже лучше называть, как правило, snakecase'ом: linked_list

Alex6Alex commented on homework05 (18.07.2022)/LinkedList.rb:
```diff
@@ -0,0 +1,91 @@
+    end
+  end
+
+  private def find_before(data)
```
Все же предпочтительнее писать приватные методы в конце класса таким образом:

```
private

def ...
end
```

Alex6Alex commented on homework05 (18.07.2022)/LinkedList.rb:
```diff
@@ -0,0 +1,91 @@
+
+  private def find_before(data)
+    node = @head
+    return false unless node.next_node
```
Думаю, что здесь тоже можно сделать возврат `nil`

Alex6Alex commented on homework05 (18.07.2022)/LinkedList.rb:
```diff
@@ -0,0 +1,91 @@
+
+  def find(data)
+    node = @head
+    return nil unless node.next_node
```
А если применить метод для пустого списка, выполнится ли метод?

zuwarka committed:
Update LinkedList.rb

Добавлена строка 38 для возвращения nil при вызове find для пустого списка
